packages/wallet/src/rollup-state-validator.ts

```diff
@@ -0,0 +1,285 @@
+      )
+      return [senderSnapshot, recipientSnapshot]
+    }
+  }
+
+  public async getTransactionFromTransitionAndSnapshots(
+    transition: RollupTransition,
+    snapshots: StateSnapshot[]
+  ): Promise<SignedTransaction> {
+    if (isTransferTransition(transition)) {
+      const sender: Address = snapshots[0].state.pubKey
+      const recipient: Address = snapshots[1].state.pubKey
+      const convertedTx: Transfer = {
+        sender,
+        recipient,
+        tokenType: transition.tokenType as UniTokenType | PigiTokenType,
+        amount: transition.amount,
+      }
+      return {
+        signature: transition.signature,
+        transaction: convertedTx,
+      }
+    } else if (isSwapTransition(transition)) {
+      const swapper: Address = snapshots[0].state.pubKey
+      const convertedTx: Swap = {
+        sender: swapper,
+        tokenType: transition.tokenType as UniTokenType | PigiTokenType,
+        inputAmount: transition.inputAmount,
+        minOutputAmount: transition.minOutputAmount,
+        timeout: transition.timeout,
+      }
+      return {
+        signature: transition.signature,
+        transaction: convertedTx,
+      }
+    }
+
+    return undefined
+  }
+
+  public async checkNextTransition(
+    nextTransition: RollupTransition
+  ): Promise<FraudCheckResult> {
+    let preppedFraudInputs: StateSnapshot[]
+    let generatedPostRoot: Buffer
+
+    const transitionPostRoot: Buffer = hexStrToBuf(nextTransition.stateRoot)
+
+    if (isCreateAndTransferTransition(nextTransition)) {
+      const slotIfSequential: number = await this.rollupMachine.getNextNewAccountSlot()
+      // if the created slot is not sequential, for now it will break
+      if (slotIfSequential < nextTransition.recipientSlotIndex) {
+        throw new AggregatorUnsupportedError()
+      }
+    }
+
+    // In case there was fraud in this transaction, get state snapshots for each input so we can prove the fraud later.
+    preppedFraudInputs = await this.getInputStateSnapshots(nextTransition)
+    // let inputAsTransaction: SignedTransaction = await this.getTransactionFromTransition(nextTransition)
+    const inputAsTransaction: SignedTransaction = await this.getTransactionFromTransitionAndSnapshots(
+      nextTransition,
+      preppedFraudInputs
+    )
+    try {
+      await this.rollupMachine.applyTransaction(inputAsTransaction)
+      generatedPostRoot = await this.rollupMachine.getStateRoot()
+    } catch (error) {
+      if (isStateTransitionError(error)) {
+        // return the fraud proof, invalid transaction
+        return {
+          fraudPosition: this.currentPosition,
+          fraudInputs: preppedFraudInputs,
+          fraudTransition: nextTransition,
+        }
+      } else {
+        throw new LocalMachineError()
+      }
+    }
+
+    if (generatedPostRoot.equals(transitionPostRoot)) {
+      this.currentPosition.blockNumber++
+      this.currentPosition.transitionIndex++
+      return 'NO_FRAUD'
```

Comment:
Would `undefined` also work? `undefined` has the benefit of being able to check fraud or not with a single `!` in an `if`.

Reply:
I thought that making sure it "indeed found no fraud" as opposed to some bug which resulted in undefined would be useful--if this is an anti-pattern, I can change.

Reply:
disregard--changing as this helps with 'getTransitionFraudProof' below